torch/csrc/jit/codegen/cuda/ir_graphviz.h

```diff
@@ -0,0 +1,121 @@
+
+#pragma once
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/csrc/jit/codegen/cuda/dispatch.h>
+
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+
+// Corresponding unit test
+void testGPU_IrGraphGenerator();
+
+namespace fuser {
+
+// Generates a DOT (https://www.graphviz.org) graph
+// representation of a fuser IR
+//
+// Usage:
+// 1) Add calls to IrGraphGenerator::print(), for example:
+//  `IrGraphGenerator::print(&fusion, "ir.dot")`
+//
+// 2) Call IrGraphGenerator::print() from a debugger. Using gdb for example:
+//  `call IrGraphGenerator::print(&fusion, "ir.dot",
+//      IrGraphGenerator::DetailLevel::Explicit)`
+//
+// Notes:
+//  - When called from the debugger, the detail_level must be
+//    explicitly passed in (most debuggers don't support default arguments)
+//
+//  - The output dot file path can't include shell specific notations,
+//    for example you can't use "~/temp/ir.dot" ("/home/user/temp/ir.dot"
+//    must be used instead)
+//
+class TORCH_CUDA_API IrGraphGenerator : private OptInConstDispatch {
+  friend void torch::jit::testGPU_IrGraphGenerator();
```

Comment:
IIUC, the only thing that we are checking in the tests are generated strings. If we slightly change the public `print` function to output string, this is not necessary any more, right?

On related topic, only being able to write output to a file seems quite awkward to me. It would be easier let user opt-in the output format. In that case we can opt for output as stringstream and test it that way. So we don't need to declare friend here.

Reply:
The intention is to make the IR printing as convenient as possible for debugging - so one call either from the code or from the debugger.

The "return string" is exactly the testing interface I had in mind, so we're probably talking about the same thing (I tried to resist adding it unless there's a good justification for it)

I'll make the changes.